### PR TITLE
Reduction fixes and gradient checks

### PR DESCRIPTION
**Pull request by AlexDBlack**
- Implements additional gradient check util methods (test all inputs for a SameDiff instance)
- Fixes for reductions backprop
- Fixes: https://github.com/deeplearning4j/nd4j/issues/2469

### CODE REVIEW

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java [hunk 1]
```diff
@@ -976,6 +975,18 @@ public int getInputLength(SDVariable func) {
         return ArrayUtil.prod(inputShape);
     }
 
+    public int getReductionLength(DifferentialFunction func){
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java]: Use ArrayUtil.prod instead?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java [hunk 2]
```diff
@@ -121,4 +127,153 @@ public static boolean checkGradients(
...
+
+
+    public static boolean checkGradients(SameDiff sd, double eps, double maxRelError, double minAbsError, boolean print,
+                                         boolean exitOnFirstFailure){
+        //Check data type:
+        if(Nd4j.dataType() != DataBuffer.Type.DOUBLE){
+            throw new IllegalStateException("Data type must be set to double");
+        }
+
+        Set<String> fnOutputs = new HashSet<>();
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java]: LinkedHashSet please

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java [hunk 3]
```diff
@@ -121,4 +127,153 @@ public static boolean checkGradients(
...
+        INDArray out = sd.execAndEndResult();
+        if(out.length() != 1){
+            throw new IllegalStateException("Output variable is not a scalar - has shape " + Arrays.toString(out.shape()));
+        }
+
+        //TODO also check that all inputs are non-zero (otherwise: consider out = sum(x * y) with all x and y being 0
+        // in this case, gradients of x and y are all 0 too
+
+        sd.execBackwards();
+        Map<String,INDArray> grad = new HashMap<>();
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java]: LinkedHashMap please

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java (hunk 2 above)
Comment (AlexDBlack) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java]: We never iterate over fnOutputs (only add and contains) so makes zero difference here.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java (hunk 3 above)
Comment (AlexDBlack) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/gradcheck/GradCheckUtil.java]: Again, never iterate over this.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java (hunk 1 above)
Comment (AlexDBlack) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java]: Line 981: ```ArrayUtil.prod(inputShape);``` and the second one can't use it (without a remove indices anyway)

**Pull request merged**